src/completions.ts

```diff
@@ -0,0 +1,127 @@
+import {
+  Completion,
+  CompletionCreateParams,
+  CompletionCreateParamsBase,
+  CompletionCreateParamsNonStreaming,
+  CompletionCreateParamsStreaming,
+  Completions,
+} from "openai/resources/completions";
+import { Stream } from "openai/streaming";
+import { LibrettoConfig, send_event } from ".";
+import { getResolvedPrompt, getResolvedStream } from "./resolvers";
+import { PiiRedactor } from "./pii";
+
+export class LibrettoCompletions extends Completions {
+  protected piiRedactor?: PiiRedactor;
+
+  constructor(
+    client: OpenAI,
+    protected config: LibrettoConfig,
+  ) {
+    super(client);
+
+    if (config.redactPii) {
+      this.piiRedactor = new PiiRedactor();
+    }
+  }
+
+  override create(
+    body: CompletionCreateParamsNonStreaming,
+    options?: Core.RequestOptions,
+  ): APIPromise<Completion>;
+  override create(
+    body: CompletionCreateParamsStreaming,
+    options?: Core.RequestOptions,
+  ): APIPromise<Stream<Completion>>;
+  override create(
+    body: CompletionCreateParamsBase,
+    options?: Core.RequestOptions,
+  ): APIPromise<Stream<Completion> | Completion>;
+  override create(
+    body: CompletionCreateParams,
+    options?: Core.RequestOptions,
+  ): APIPromise<Completion> | APIPromise<Stream<Completion>> {
+    return this._create(body, options) as
+      | APIPromise<Completion>
+      | APIPromise<Stream<Completion>>;
+  }
+
+  private async _create(
+    body: CompletionCreateParams,
+    options?: Core.RequestOptions,
+  ): Promise<Completion | Stream<Completion>> {
+    const now = Date.now();
+    const { libretto, prompt, stream, ...openaiBody } = body;
+
+    const { prompt: resolvedPrompt, template } = getResolvedPrompt(
+      prompt,
+      libretto?.templateParams,
+    );
+
+    const resultPromise = super.create(
+      { ...openaiBody, prompt: resolvedPrompt, stream },
+      options,
+    ) as APIPromise<Completion> | APIPromise<Stream<Completion>>;
+
+    const resolvedPromptStr = Array.isArray(resolvedPrompt)
+      ? null
+      : resolvedPrompt;
+
+    const resolvedPromptTemplateName =
+      libretto?.promptTemplateName ?? this.config.promptTemplateName;
+
+    if (!resolvedPromptTemplateName && !this.config.allowUnnamedPrompts) {
+      return resultPromise;
+    }
+
+    const feedbackKey = libretto?.feedbackKey ?? crypto.randomUUID();
+
+    const { finalResultPromise, returnValue } = await getResolvedStream(
+      resultPromise,
+      stream,
+      feedbackKey,
+      false,
+    );
+    finalResultPromise.then((response) => {
+      const responseTime = Date.now() - now;
+      let params = libretto?.templateParams ?? {};
+
+      // Redact PII before recording the event
+      if (this.piiRedactor) {
+        try {
+          response = this.piiRedactor.redact(response);
+          params = this.piiRedactor.redact(params);
+        } catch (err) {
+          console.log("Failed to redact PII", err);
+        }
+      }
+
+      send_event({
```

Comment:
since we're not awaiting the outer promise, I think we *should* await this here, so that it's rolled up into one unresolved promise instead of 2. (here and the other call)

In the future, I think we'll want to accumulate these unawaited promises and add some kind of awaitOutstandingEvents() or something